x16-smc.ino

```diff
@@ -572,13 +641,35 @@ void mouseClockIrq() {
 // ----------------------------------------------------------------
 // Bootloader Startup
 // ----------------------------------------------------------------
-void startBootloader() {
-  if (bootloaderFlags == 3 && bootloaderTimer > 0) {
-    ((void(*)(void))BOOTLOADER_START_ADDR)();
-  }
-  else {
-    bootloaderTimer = 50;
+void evaluateBootloaderKeypress() {
+  if (bootloaderTimer > 0 && specialModeActive != 1) {
+    uint8_t temp = SREG;
+    cli();
+    bootloaderTimer--;
+    SREG = temp;
```

Comment:
I'm not sure I follow the exact logic of disabling interrupts during this sequence of code.

I understand that startBootloader is called from the main loop, and could be interrupted.

The value of bootloaderTimer could be changed within in an interrupt before the remainder of this function has been executed.

Reply:
It is because the 16 bit volatile variable bootloaderTimer is decremented in main loop while it can be set elsewhere (likely in interrupts). Given the code

```
    bootloaderTimer--;
    15ca:	80 91 53 01 	lds	r24, 0x0153	; 0x800153 <bootloaderTimer>
    15ce:	90 91 54 01 	lds	r25, 0x0154	; 0x800154 <bootloaderTimer+0x1>
    15d2:	01 97       	sbiw	r24, 0x01	; 1
    15d4:	90 93 54 01 	sts	0x0154, r25	; 0x800154 <bootloaderTimer+0x1>
    15d8:	80 93 53 01 	sts	0x0153, r24	; 0x800153 <bootloaderTimer>
```

, 15d4 writes the high value back to ram. 15d8 writes the low value back to ram. There can trigger an interrupt between 15d4 and 15d8 changing both places in ram. When 15d8 executes after the interrupt, the value of bootloaderTimer will not contain an intended value. To ensure that the value of bootloaderTimer is always an intended value, I do the decrement as an atomic operation. (Perhaps I should use ATOMIC_BLOCK_RESTORE instead to make the intent more clear). Also, there can trigger an interrupt between 15ca and 15ce, also causing that wrong number gets decremented.

Reply:
OK, that makes sense.